### PR TITLE
Fix a Nested(many=True).dump swallowing a first element from a generator

### LINKED ISSUE #1163: 2.x: Nested(many=True) eats first element from generator value when dumping
As reproduced in Python 3.6.8:

```py
from marshmallow import Schema, fields

class O(Schema):
    i = fields.Int()

class P(Schema):
    os = fields.Nested(O, many=True)

def gen():
    yield {'i': 1}
    yield {'i': 0}

p = P()
p.dump({'os': gen()})
# MarshalResult(data={'os': [{'i': 0}]}, errors={})
```

Problematic code is here:

https://github.com/marshmallow-code/marshmallow/blob/2.x-line/src/marshmallow/fields.py#L447

And here:

https://github.com/marshmallow-code/marshmallow/blob/2.x-line/src/marshmallow/schema.py#L832

The easiest solution would be to cast `nested_obj` to list before calling `schema._update_fields`, just like a normal Schema with `many=True` does.

### PR DESCRIPTION
**Pull request by khvn26**
Fixes #1163

### CODE REVIEW

**Review by sloria — CHANGES_REQUESTED** [review]
I think this will require some more investigation. It is common to serialize non-list objects like querysets, where it would be undesirable to cast to a list.

**Comment by khvn26** [comment]
> It is common to serialize non-list objects like querysets, where it would be undesirable to cast to a list.

Aren't they already cast to list though if a schema is not nested? See https://github.com/marshmallow-code/marshmallow/blob/2.x-line/src/marshmallow/schema.py#L489

**Comment by sloria** [comment]
Ah, that's right. I'm still curious why the casting in `Nested` isn't necessary on `dev`. I don't have time to look into it right now, but hopefully over the weekend.

**Comment by deckar01** [comment]
TLDR; :+1:

It looks like `list(obj)` was previously only used to cast generators, but it was changed to all iterables due to #353. #353 is basically the same bug as #1163, but at the top level. See #355

`itertools.tee()` was proposed as an alternate solution, but the performance impact was never discussed. Peeking at the first item isn't exactly what that utility was designed to do and based on the docs it will probably use a little more memory than is necessary. There are simpler ways to just peek at the first item. See https://stackoverflow.com/questions/2425270/how-to-look-ahead-one-element-in-a-python-generator

That said, internally marshmallow will eventually use a list comprehension that blocks on the generator. I replaced the list cast with a simple peek and found no significant change in performance. I replaced the list comprehension in the marshaller with a generator expression and found that for simple schemas, marshmallow would return a generator practically instantly. I think this was the desired behavior.

This seems like an acceptable fix for 2.x-line since the marshaller already blocks and reads everything into memory. We might want to explore streaming with generators in 3.x.

**Comment by sloria** [comment]
Oh, I see now. The casting isn't necessary on dev because of the #810, which obviated the indexing of the first element. OK, I think this is fine for now.

**Review by sloria — APPROVED** [review]

**Comment by khvn26** [comment]
Thank you! Now I'm obligated to ask when the next release is :)

**Comment by sloria** [comment]
This is released in 2.19.1.
